### PR TITLE
[CCB] Dsjson and json parsing for CCB

### PR DESCRIPTION
**Pull request by jackgerrits**
- Implement JSON parsing in normal and dsjson mode for CCB
- Add tests for dsjson and json for simple, cb and ccb

Note: this will not build until #1761 is merged

### CODE REVIEW

File: vowpalwabbit/parse_example_json.h [hunk 1]
```diff
@@ -640,16 +751,40 @@ class DefaultState : public BaseState<audit>
       if (ctx.key_length == 6 && !strcmp(str, "_multi"))
         return &ctx.multi_state;
 
+      if (ctx.key_length == 3 && !strcmp(str, "_df"))
+        return &ctx.df_state;
+
       if (ctx.key_length == 4 && !_stricmp(ctx.key, "_tag"))
         return &ctx.tag_state;
 
+      if (ctx.key_length == 4 && !_stricmp(ctx.key, "_inc"))
```
Comment (ataymano) [inline vowpalwabbit/parse_example_json.h]: is there any specific reason for using ctx.key in some places and str in others?

File: vowpalwabbit/parser.cc [hunk 2]
```diff
@@ -549,19 +549,33 @@ void enable_sources(vw& all, bool quiet, size_t passes, input_options& input_opt
...
+        }
+
         // TODO: change to class with virtual method
         // --invert_hash requires the audit parser version to save the extra information.
         if (all.audit || all.hash_inv)
         {
           all.p->reader = &read_features_json<true>;
           all.p->audit = true;
-          all.p->jsonp = std::make_shared<json_parser<true>>();
+          auto parser_context = std::make_shared<json_parser<true>>();
```
Comment (JohnLangford) [inline vowpalwabbit/parser.cc]: Minor, but why don't you assign to the final destination and then modify?

Comment (JohnLangford) [reply vowpalwabbit/parser.cc]: Even better, should mode be an argument to the constructor?

File: vowpalwabbit/parse_example_json.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/parse_example_json.h]: I don't believe there is a reason, especially since it is set at the top of the function

**Comment by JohnLangford** [comment]
Maybe add a test an end-to-end test in RunTests?  That will document the usage pattern.

**Pull request merged**